node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java

```diff
@@ -0,0 +1,442 @@
+        var nodeMap = constructHostnameToNodeMap(hosts);
+        var nodeChildren = constructNodeChildrenMap(tenants, hosts, nodeMap);
+        var availableResources = constructAvailableResourcesMap(hosts, nodeChildren);
+
+        List<Node> overcommittedNodes = findOvercommittedNodes(availableResources);
+        if (overcommittedNodes.size() != 0) {
+            log.log(LogLevel.WARNING, String.format("%d nodes are overcommitted! [ %s ]", overcommittedNodes.size(),
+                    overcommittedNodes.stream().map(Node::hostname).collect(Collectors.joining(", "))));
+        }
+        return overcommittedNodes.size();
+    }
+
+    private Map<String, Node> constructHostnameToNodeMap(List<Node> nodes) {
+        return nodes.stream().collect(Collectors.toMap(Node::hostname, n -> n));
+    }
+
+    private Map<Node, List<Node>> constructNodeChildrenMap(List<Node> tenants, List<Node> hosts, Map<String, Node> hostnameToNode) {
+        Map<Optional<Node>, List<Node>> possibleNodeChildren = tenants.stream()
+                .filter(n -> n.parentHostname().isPresent())
+                .collect(Collectors.groupingBy(
+                        n -> Optional.ofNullable(hostnameToNode.get(n.parentHostname().orElseThrow()))));
+        possibleNodeChildren.remove(Optional.<Node>empty());
+        Map<Node, List<Node>> nodeChildren = possibleNodeChildren.entrySet().stream()
+                .collect(Collectors.toMap(e -> e.getKey().orElseThrow(), Map.Entry::getValue));
+
+        for (var host : hosts) nodeChildren.putIfAbsent(host, List.of());
+
+        return nodeChildren;
+    }
+
+    private Map<Node, AllocationResources> constructAvailableResourcesMap(List<Node> hosts, Map<Node, List<Node>> nodeChildren) {
+        Map<Node, AllocationResources> availableResources = new HashMap<>();
+        for (var host : hosts) {
+            NodeResources hostResources = host.flavor().resources();
+            int occupiedIps = 0;
+            Set<String> ipPool = host.ipAddressPool().asSet();
+            for (var child : nodeChildren.get(host)) {
+                hostResources = hostResources.subtract(child.flavor().resources());
+                occupiedIps += child.ipAddresses().stream().filter(ipPool::contains).count();
+            }
+            availableResources.put(host, new AllocationResources(hostResources, host.ipAddressPool().asSet().size() - occupiedIps));
+        }
+
+        return availableResources;
+    }
+
+    /**
+     * Computes a heuristic for each host, with a lower score indicating a higher perceived likelihood that removing
+     * the host causes an unrecoverable state
+     */
+    private Map<Node, Integer> computeMaximalRepeatedRemovals(List<Node> hosts, Map<Node, List<Node>> nodeChildren,
+                                                      Map<Node, AllocationResources> availableResources) {
+        Map<Node, Integer> timesNodeCanBeRemoved = hosts.stream().collect(Collectors.toMap(
```

Comment:
Because of the bit which I forgot to put in where I skip childless hosts, thus necessitating initializing them at max_int so they'll be sorted to the bottom.
```Java
        Map<Node, Integer> timesNodeCanBeRemoved = hosts.stream().collect(Collectors.toMap(
                Function.identity(),
                _x -> Integer.MAX_VALUE
        ));
        for (Node host : hosts) {
            List<Node> children = nodeChildren.get(host);
            if (children.size() == 0) continue;
```
Thanks for pointing it out!